core/src/main/java/io/parsingdata/metal/Util.java

```diff
@@ -111,6 +118,33 @@ public Optional<Value> eval(final Value value, final ParseState parseState, fina
         };
     }
 
+    public static ValueExpression digest(final String algorithm, final ValueExpression target) {
+        return new UnaryValueExpression(target) {
+            @Override
+            public Optional<Value> eval(final Value value, final ParseState parseState, final Encoding encoding) {
+                try {
+                    final Slice slice = value.slice();
+                    final BigInteger bufferSize = valueOf(512);
+                    final MessageDigest digester = getInstance(algorithm);
+
+                    for (BigInteger i = ZERO; i.compareTo(slice.length) < 0; i = i.add(bufferSize)) {
+                        final BigInteger length = bufferSize.min(slice.length.subtract(i));
+                        final Optional<Slice> subslice = createFromSource(slice.source, slice.offset.add(i), length);
```

Comment:
Instead of calculating the exact slice you can also create a slice where only the offset is moved.
I.e.:
Your implementation uses this as slices: 
```
|***********|
|***|
   |***|
      |***|
         |**|
```
But what if we create slices as:
```
|***********|
|***********|
   |********|
      |*****|
         |**|
```
In combination with the `getData(final int limit)` method on `slice`, the calculations would be a easier.

Reply:
Can you give a code example? I don't know how to do this. `createFromSource` is the only public way to create new slices, right?

Reply:
@mvanaken

Reply:
I was thinking about something like this:
```java
Slice slice = value.slice();
while (slice != null) {
    digester.update(slice.getData(bufferSize));
    slice = createFromSource(slice.source, slice.offset.add(bufferSize), slice.length.subtract(bufferSize)).orElse(null);
}
return Optional.of(new CoreValue(createFromBytes(encoding.byteOrder.apply(digester.digest())), encoding));
```
if you want to differentiate between "end of slicing" and "slice should exist but isn't", you could extend it with:
```java
Slice slice = value.slice();
while (slice != null && slice.length.compareTo(bufferSize) > 0) {
    digester.update(slice.getData(bufferSize));
    slice = createFromSource(slice.source, slice.offset.add(bufferSize), slice.length.subtract(bufferSize)).orElse(null);
}
if (slice == null) {
    return Optional.empty();
}
digester.update(slice.getData(bufferSize));
return Optional.of(new CoreValue(createFromBytes(encoding.byteOrder.apply(digester.digest())), encoding));
```
This is all untested, so I may have missed something.
There is not much difference between the second option and your code, but it does avoid the Optional.get() call.